src/chunk_scan.c

```diff
@@ -61,105 +59,92 @@ ts_chunk_scan_by_chunk_ids(const Hyperspace *hs, const List *chunk_ids, unsigned
 	foreach (lc, chunk_ids)
 	{
 		int chunk_id = lfirst_int(lc);
-		TupleInfo *ti;
 
 		Assert(CurrentMemoryContext == work_mcxt);
 
 		ts_chunk_scan_iterator_set_chunk_id(&chunk_it, chunk_id);
 		ts_scan_iterator_start_or_restart_scan(&chunk_it);
-		ti = ts_scan_iterator_next(&chunk_it);
-
-		if (ti)
+		TupleInfo *ti = ts_scan_iterator_next(&chunk_it);
+		if (ti == NULL)
 		{
-			bool isnull;
-			Datum datum = slot_getattr(ti->slot, Anum_chunk_dropped, &isnull);
-			bool is_dropped = isnull ? false : DatumGetBool(datum);
-
-			MemoryContextSwitchTo(per_tuple_mcxt);
-			MemoryContextReset(per_tuple_mcxt);
-
-			if (!is_dropped)
-			{
-				Chunk *chunk = MemoryContextAllocZero(orig_mcxt, sizeof(Chunk));
+			continue;
+		}
+		bool isnull;
+		Datum datum = slot_getattr(ti->slot, Anum_chunk_dropped, &isnull);
+		const bool is_dropped = isnull ? false : DatumGetBool(datum);
+		if (is_dropped)
+		{
+			continue;
+		}
 
-				MemoryContext old_mcxt = MemoryContextSwitchTo(ti->mctx);
-				ts_chunk_formdata_fill(&chunk->fd, ti);
-				MemoryContextSwitchTo(old_mcxt);
+		/* We found a chunk that is not dropped. First, try to lock it. */
+		Name schema_name = DatumGetName(slot_getattr(ti->slot, Anum_chunk_schema_name, &isnull));
+		Assert(!isnull);
+		Name table_name = DatumGetName(slot_getattr(ti->slot, Anum_chunk_table_name, &isnull));
+		Assert(!isnull);
 
-				chunk->constraints = NULL;
-				chunk->cube = NULL;
-				chunk->hypertable_relid = hs->main_table_relid;
+		Oid chunk_reloid = ts_get_relation_relid(NameStr(*schema_name),
+												 NameStr(*table_name),
+												 /* return_invalid = */ false);
+		Assert(OidIsValid(chunk_reloid));
 
-				unlocked_chunks[unlocked_chunk_count] = chunk;
-				unlocked_chunk_count++;
-			}
+		/* Only one chunk should match */
+		Assert(ts_scan_iterator_next(&chunk_it) == NULL);
 
-			MemoryContextSwitchTo(work_mcxt);
-			/* Only one chunk should match */
-			Assert(ts_scan_iterator_next(&chunk_it) == NULL);
+		DEBUG_WAITPOINT("hypertable_expansion_before_lock_chunk");
+		if (!ts_chunk_lock_if_exists(chunk_reloid, AccessShareLock))
+		{
+			continue;
 		}
-	}
 
-	ts_scan_iterator_close(&chunk_it);
+		/*
+		 * Now after we have locked the chunk, we have to reread its metadata.
+		 * It might have been modified concurrently by decompression, for
+		 * example.
+		 */
+		ts_chunk_scan_iterator_set_chunk_id(&chunk_it, chunk_id);
+		ts_scan_iterator_start_or_restart_scan(&chunk_it);
+		ti = ts_scan_iterator_next(&chunk_it);
+		Assert(ti != NULL);
+		Chunk *chunk = MemoryContextAllocZero(orig_mcxt, sizeof(Chunk));
 
-	Assert(unlocked_chunk_count == 0 || unlocked_chunks != NULL);
-	Assert(unlocked_chunk_count <= list_length(chunk_ids));
-	Assert(CurrentMemoryContext == work_mcxt);
+		MemoryContextSwitchTo(per_tuple_mcxt);
+		MemoryContextReset(per_tuple_mcxt);
 
-	DEBUG_WAITPOINT("expanded_chunks");
+		MemoryContext old_mcxt = MemoryContextSwitchTo(ti->mctx);
```

Comment:
If I understand the code flow correctly, the `old_mcxt` will always be the `per_tuple_mcxt` at this point. Is this intended and couldn't we perform the reset of this context after this line and prevent the MemoryContext switch in line 115?

Reply:
I removed the `per_tuple_mcxt` altogether, I think two separate temporary memory contexts are an overkill for this function.